examples/image_captioning/datasets.py

```diff
@@ -0,0 +1,138 @@
+from collections import defaultdict
+import os
+
+import numpy as np
+from PIL import Image
+from pycocotools.coco import COCO
+
+from chainer import dataset
+from chainer.dataset.convert import to_device
+
+
+# Vocabulary tokens of BOS (beginning of sentence), EOS (end of sentence),
+# UNK (unknown word) and token labels to be ignored in the loss computation by
+# the LSTM layers
+_bos = 0
+_eos = 1
+_unk = 2
+_ignore = -1
+
+
+def split(sentence):
+    return sentence.lower().replace('.', ' .').replace(',', ' ,').split()
+
+
+class MsCocoDataset(dataset.DatasetMixin):
+
+    """Wraps the MSCOCO datasets and is used by the iterator to fetch data."""
+
+    def __init__(self, root_dir, data_dir, anno_file):
+        coco = COCO(os.path.join(root_dir, anno_file))
+        anns = coco.loadAnns(coco.getAnnIds())
+
+        self.coco = coco
+        self.anns = anns
+        self.vocab = None  # Later set from outside
+        self.coco_root = root_dir
+        self.coco_data = data_dir
+
+    def __len__(self):
+        return len(self.anns)
+
+    def get_example(self, i):
+        """Called by the iterator to fetch a data sample.
+
+        A data sample from MSCOCO consists of an image and its corresponding
+        caption.
```

Comment:
I think it is nice if the shape of returned image is described here.

Reply:
Fixed.